api/grpcserver/post_client.go

```diff
@@ -0,0 +1,101 @@
+package grpcserver
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	pb "github.com/spacemeshos/api/release/go/spacemesh/v1"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+)
+
+// postClient represents a connection to a PoST service.
+//
+// It uses the grpc interface of the node to send commands to the post service.
+// Additionally if instructed it will start the post service and connect it to
+// the node.
+type postClient struct {
+	service *PostService
+
+	closed chan struct{}
+}
+
+func newPostClient(service *PostService) *postClient {
+	return &postClient{
+		service: service,
+		closed:  make(chan struct{}),
+	}
+}
+
+func (pc *postClient) Proof(ctx context.Context, challenge []byte) (*types.Post, *types.PostMetadata, error) {
+	ch := make(chan *pb.ServiceResponse, 1)
+	cmd := postCommand{
+		req: &pb.NodeRequest{
+			Kind: &pb.NodeRequest_GenProof{
+				GenProof: &pb.GenProofRequest{
+					Challenge: challenge,
+				},
+			},
+		},
+		ch: ch,
+	}
+
+	select {
+	case <-pc.closed:
+		return nil, nil, fmt.Errorf("post client closed")
+	default:
+	}
```

Comment:
I was primarily concerned with giving a concrete error when trying to send a request when the client has already been closed.

If `Proof` is called when `pc.closed` has already been closed it will return the error `post client closed`. Should the close happen right between this check and the following `pc.service.Send` it will return `failed to send request: connection not established`.

Without this check the client would send to a (possibly) different Post service as soon as connection is re-established. I know this can theoretically still happen, but for this the current connection has to be dropped and a new connection established between here and `pc.service.Send` which I don't think is enough time to happen?